File: src/converter/plugins/ConfigurationPlugin.ts
import {ConverterPlugin} from "../plugin.js";
import {Node} from "typescript";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {Configuration} from "../../configuration/configuration.js";
import {GeneratedFile} from "../generated.js";

export const configurationServiceKey = Symbol()

export class ConfigurationService {
    constructor(readonly configuration: Configuration) {
    }
}

export class ConfigurationPlugin implements ConverterPlugin {
    private readonly configurationService: ConfigurationService;

    constructor(configuration: Configuration) {
        this.configurationService = new ConfigurationService(configuration);
    }

    generate(): GeneratedFile[] {
        return [];
    }

    render(node: Node, context: ConverterContext, next: Render): string | null {
        return null;
    }

    traverse(node: Node): void {
    }

    setup(context: ConverterContext): void {
        context.registerService(configurationServiceKey, this.configurationService)
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/ConfigurationPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.extension.ConverterContext
import io.github.sgrishchenko.karakum.extension.ConverterPlugin
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Render
import js.symbol.Symbol
import typescript.Node

@OptIn(ExperimentalJsExport::class)
@JsExport
val configurationServiceKey = Symbol()

@OptIn(ExperimentalJsExport::class)
@JsExport
class ConfigurationService @JsExport.Ignore constructor(val configuration: Configuration)

class ConfigurationPlugin(configuration: Configuration) : ConverterPlugin<Node> {
    private val configurationService = ConfigurationService(configuration)

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun render(node: Node, context: ConverterContext, next: Render<Node>) = null

    override fun generate(context: ConverterContext, render: Render<Node>) = emptyArray<GeneratedFile>()

    override fun setup(context: ConverterContext) {
        context.registerService(configurationServiceKey, configurationService)
    }
}
